lib/Runtime/Math/JavascriptMath.cpp

```diff
@@ -227,6 +227,15 @@ using namespace Js;
             Js::TypeId typeLeft = JavascriptOperators::GetTypeId(aLeft);
             Js::TypeId typeRight = JavascriptOperators::GetTypeId(aRight);
 
+            if (typeLeft == TypeIds_BigInt || typeRight == TypeIds_BigInt)
```

Comment:
Could this go in JavascriptMath::Add_FullHelper right before the call to Add_Helper? That way you don't have to duplicate it because both of these methods end up calling that helper function. Also, I expect the profiling compiler will already do this for us, but ideally we would check for more common cases (numbers and strings) first.

Reply:
@sethbrenith I need to duplicate it here because we have 2 different functions for x64 and x86. I think I can put it in inline JavascriptMath::Add but it is "inline" so I am not sure if I should.

Reply:
There are two different methods here based on `#if FLOATVAR`, but they both end up calling Add_FullHelper_Wrapper, which calls Add_FullHelper. So I think Add_FullHelper is a reasonable place to add this so we only have to do it once.

Also, I just realized there's also a correctness concern. We're supposed to call ToPrimitive on the inputs, which Add_FullHelper_Wrapper does. Consider: `1n + {valueOf: () => 2n}`. According to the spec we're supposed to coerce the second arg to a primitive, which calls its `valueOf` method. I think your current implementation would throw. This is also a problem for subtraction.

So basically, I think the right way to think about this method is that it's optimized fast paths for some common cases, but the more complete code belongs in Add_FullHelper.

---
In reply to: [227879760](https://github.com/Microsoft/ChakraCore/pull/5792#discussion_r227879760) [](ancestors = 227879760)

Reply:
You don't need to implement valueOf; that's part of ToPrimitive. The change you made should in ToPrmitive is enough; I just verified on your latest iteration that `1n + {valueOf: () => 2n} == 3n`.

Reply:
I'm happy to add this to test :)